cmd/dagger/engine.go

```
@@ -106,7 +107,17 @@ func withEngine(
 
 			if exist {
 				for _, dep := range mod.Dependencies {
-					err := dep.AsModule().Serve(ctx)
+					depKind, err := dep.Kind(ctx)
+					if err != nil {
+						return fmt.Errorf("failed to get dependency kind: %w", err)
+					}
+
+					// Skip remote dependencies because they are already serves during the call to `Connect`
+					if depKind == dagger.ModuleSourceKindGitSource {
+						continue
+					}
```

Hmm, this seems a little weird to me. I'm curious why `Connect` can't also load the local deps - it's still the current directory right?

Makes more sense to keep the logic in the same place right? Or is there a reason why it can't be unified?

---

It's because the local dependencies may not be available depending on where the code is run.
We would also need to deal with local path resolution while this is already handle by the engine.

So to avoid complexities, I just choose to embed remote dependency loading.

---

> because the local dependencies may not be available depending on where the code is run

Yeah, but this is the same even if we embed in `dagger run` where the logic is now right?

---

Yeah but then it's going to fail if the local dependency is called, the server will throw an error to mention that the object doesn't exist, so it gracefully fail IMO

---

Do you have a flexible solution to deal with path relative to the generated client? Because if the client is generated in `a/b/c` and the dagger.json is in `a` and a local deps is in `../d`, then if you move the client file, everything will break :/

---

I think it's *really* confusing that different dependencies are loaded differently, where local ones are loaded in an entirely different part of the code base from the remote ones.

> Last wednesday we agreed to only embed remote dependency, so isn't it fine enough for now?

Yeah, we did, but we're still handling local dependencies here - it's not like we're just ignoring those, we're still trying to load them?

> Yes but we do not want to error in case of local dependency embedding because if you create a CLI or something, the user will not have access to the local dep

Right. Yes. But IMO, better an explicit error if we hit this case, rather than skipping the loading of a dependency that we've tried to install - which then fails when you try to use it, right?

FYI @sipsma @helderco

---

> Right. Yes. But IMO, better an explicit error if we hit this case, rather than skipping the loading of a dependency that we've tried to install - which then fails when you try to use it, right?

Since local dependency are mostly for local development, this situation would happen if a local dep is used for a published library :/
I'll add that changes because it makes sense but I'm scared that it will be quite unstable regarding path resolutions

---

Ohhhh I did a try to add local deps and that's way harder than expected because it depends where the `main`  is executed... so if your main command is in a sub directory, the relative path to that local module is different.
I'm not sure how to tackle that :/